packages/libs/error-stack/src/fmt/hook.rs

```diff
@@ -0,0 +1,591 @@
+// we could do cfg for everything, but that gets very messy, instead we only use a subset
+// and enable deadcode on `feature = "std"`.
+#![cfg_attr(not(feature = "std"), allow(dead_code))]
+
+use alloc::{boxed::Box, collections::BTreeMap, string::String, vec::Vec};
+use core::{
+    any::{Any, TypeId},
+    marker::PhantomData,
+};
+
+pub use default::builtin;
+
+use crate::fmt::{Emit, Frame};
+
+#[derive(Default)]
+pub struct HookContextImpl {
+    pub(crate) snippets: Vec<String>,
+    alternate: bool,
+
+    storage: BTreeMap<TypeId, BTreeMap<TypeId, Box<dyn Any>>>,
+}
+
+impl HookContextImpl {
+    pub(crate) fn as_hook_context<T>(&mut self) -> HookContext<'_, T> {
+        HookContext {
+            parent: self,
+            _marker: PhantomData::default(),
+        }
+    }
+
+    pub(crate) const fn alternate(&self) -> bool {
+        self.alternate
+    }
+}
+
+/// Optional context used to carry information across hook invocations.
+///
+/// `HookContext` has two fundamental use-cases:
+/// 1) Emitting Snippets
+/// 2) Storage
+///
+/// ## Emitting Snippets
+///
+/// A [`Debug`] backtrace consists of two different sections, a rendered tree of objects and
+/// additional text/information that is too large to fit into the tree.
+///
+/// Snippets can be added to the current output via [`attach_snippet()`].
+///
+/// [`attach_snippet()`]: HookContext::attach_snippet
+/// [`Debug`]: core::fmt::Debug
+///
+/// ### Example
+///
+/// ```rust
+/// # // we only test on nightly, therefore report is unused (so is render)
+/// # #![cfg_attr(not(nightly), allow(dead_code, unused_variables, unused_imports))]
+/// use std::io::ErrorKind;
+///
+/// use error_stack::{fmt::{Emit}, Report};
+///
+/// Report::install_debug_hook::<u64>(|val, ctx| {
+///     if ctx.alternate() {
+///         ctx.attach_snippet("u64 has been encountered");
+///     }
+///     
+///     Emit::next(val.to_string())
+/// });
+///
+/// let report = Report::new(std::io::Error::from(ErrorKind::InvalidInput))
+///     .attach(2u64)
+///     .attach(3u64);
+///
+/// # owo_colors::set_override(true);
+/// # fn render(value: String) -> String {
+/// #     let backtrace = regex::Regex::new(r"Backtrace No\. (\d+)\n(?:  .*\n)*  .*").unwrap();
+/// #     let backtrace_info = regex::Regex::new(r"backtrace with (\d+) frames \((\d+)\)").unwrap();
+/// #
+/// #     let value = backtrace.replace_all(&value, "Backtrace No. $1\n  [redacted]");
+/// #     let value = backtrace_info.replace_all(value.as_ref(), "backtrace with [n] frames ($2)");
+/// #
+/// #     ansi_to_html::convert_escaped(value.as_ref()).unwrap()
+/// # }
+/// #
+/// # #[cfg(nightly)]
+/// # expect_test::expect_file![concat!(env!("CARGO_MANIFEST_DIR"), "/tests/snapshots/doc/fmt__hookcontext_emit.snap")].assert_eq(&render(format!("{report:#?}")));
+/// #
+/// println!("{report:#?}");
+/// ```
+///
+/// <pre>
+#[doc = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/snapshots/doc/fmt__hookcontext_emit.snap"))]
```

Comment:
The output does not match the code above:
```
invalid input parameter
├╴src/fmt/hook.rs:18:14
├╴2
├╴3
╰╴backtrace with [n] frames (1)

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

Backtrace No. 1
  [redacted]
```

I would have expected the snippets added

Reply:
fixed in 93a150aa4c1193cae65f69f3e877034c1faf63e9